File: src/sim/vent.ts
// Hydrothermal vent: a point feature embedded in the seafloor rock
// that erupts on a slow schedule, emitting hot particles upward and
// raising the local temperature while active. The vent is a *true
// source* in the mass ledger -- emitted mass is recorded on
// world.ventEmitted so worldMass() can still close.
//
// Schedule (sim seconds, world.dayPeriod defaults to 90):
//   - Dormant for ~2 in-game days +/- jitter between eruptions.
//   - Warmup (intensity 0 -> 1) over WARMUP_SEC.
//   - Main phase (intensity = 1) for MAIN_SEC.
//   - Cooldown (intensity 1 -> 0) over COOLDOWN_SEC.
// Total eruption visible window is ~12-15 sim seconds.
//
// Materials: 8-10 chems, weighted common-vs-rare. Selection draws
// once per emitted particle from a cumulative-weight table.

import type { World, VentState } from "./core";
import type { Molecules } from "./chem-ids";
import {
  MOLECULE_INDEX, MOLECULE_IDS, emptyMolecules,
  CHEM_CO2, CHEM_MIN, CHEM_WASTE, CHEM_AA, CHEM_FA, CHEM_GLU,
  CHEM_O2, CHEM_BIOPOLYMER, CHEM_MEMBRANE, CHEM_MARKER0,
} from "./chem-ids";

// Schedule constants. Two game days at dayPeriod=90 = 180 sim seconds
// between eruptions; small symmetric jitter so the rhythm is
// recognizable but not perfectly clockwork.
export const VENT_CYCLE_DAYS = 2;
const VENT_CYCLE_JITTER_FRAC = 0.15; // +/-15% on the dormant span
const VENT_WARMUP_SEC = 2.0;
const VENT_MAIN_SEC = 8.0;
const VENT_COOLDOWN_SEC = 3.0;

// Emission cadence while active: one batch per EMIT_PERIOD sim seconds.
const VENT_EMIT_PERIOD = 0.10;
// Expected particles per emission batch at intensity=1. Scales by
// intensity^2 so warmup/cooldown shed dramatically fewer particles
// than the peak. Fractional values << 1 are handled by floor(rate + rng):
// most batches emit 0 particles, occasional batches emit 1, long-run
// average is the rate. Dialed down 1000x further (4 -> 0.04 -> 0.0004):
// the vent is now a rare-puff feature instead of a continuous plume,
// since heavier output dominated the ledger in long runs.
const VENT_BATCH_AT_PEAK = 0.0004;

// Upward emission velocity. Particles get a strong vy < 0 (UP in y-down
// coords) plus lateral spread; gravity + buoyancy take over once they
// clear the rock mouth.
const VENT_EXIT_SPEED = 90;
const VENT_LATERAL_SPREAD = 35;
const VENT_DEPTH_SPREAD = 6;

// Temperature contribution. Active vent raises local water temperature
// by up to VENT_TEMP_PEAK degrees inside a soft Gaussian bubble of
// radius VENT_TEMP_RADIUS. temperatureAt in sim.ts reads ventHeatAt.
export const VENT_TEMP_PEAK = 40;
export const VENT_TEMP_RADIUS = 90;

// Eruption material spec: chemId + weight. Common (mineral / CO2 /
// waste) dominate; rare (amino acids, fatty acids, glucose, marker0)
// are present but sparse so they're a real event when emitted. Total
// weight is open-ended; selection normalizes on the fly.
interface EmitSpec {
  chemId: number;
  weight: number;
  // Density for the emitted particle. Vent precipitates are dense
  // (sulfides, metal salts -> heavy minerals); organics ride on
  // buoyancy from the upward velocity rather than density alone.
  density: number;
  // Mass per emitted particle. Sets the granularity of the ledger;
  // 1 unit per particle keeps it simple and proportional to count.
  unitMass: number;
}

const VENT_EMISSIONS: EmitSpec[] = [
  // Common: classic vent chemistry. Hot sulfide-rich brine carrying
  // dissolved CO2 and precipitating mineral grains.
  { chemId: CHEM_MIN, weight: 7.0, density: 1.8, unitMass: 1 },
  { chemId: CHEM_CO2, weight: 5.0, density: 0.6, unitMass: 1 },
  { chemId: CHEM_WASTE, weight: 3.0, density: 1.1, unitMass: 1 },
  // Slightly less common: dissolved O2 (residual from photochemistry
  // higher up in the column getting recycled through hot rock).
  { chemId: CHEM_O2, weight: 1.5, density: 0.4, unitMass: 1 },
  // Rare: abiogenic precursors. Hydrothermal vents are a leading
  // theory for the origin of organic chemistry; emit these at low
  // weight so they read as a real find when they appear in the wild.
  { chemId: CHEM_AA, weight: 1.0, density: 1.0, unitMass: 1 },
  { chemId: CHEM_FA, weight: 1.0, density: 0.9, unitMass: 1 },
  { chemId: CHEM_GLU, weight: 0.7, density: 1.0, unitMass: 1 },
  { chemId: CHEM_BIOPOLYMER, weight: 0.6, density: 1.0, unitMass: 1 },
  { chemId: CHEM_MEMBRANE, weight: 0.4, density: 0.95, unitMass: 1 },
  // Very rare: exotic marker chemistry -- gives evolution a sniffable
  // tracer chemical that only originates near the vent.
  { chemId: CHEM_MARKER0, weight: 0.2, density: 1.0, unitMass: 1 },
];

// Pre-resolved chemId -> molecule-index map for ledger updates. Built
// once at module load so the emission hot path is one array lookup.
const VENT_EMIT_MOL_IDX: number[] = (() => {
  const idx: number[] = [];
  // chem-ids.ts NAMED_CHEMICALS layout matches chemId == named slot id.
  // The molecule index for each name is in MOLECULE_INDEX.
  const CHEM_TO_KEY: Record<number, keyof Molecules> = {
    [CHEM_MIN]: "minerals",
    [CHEM_CO2]: "co2",
    [CHEM_WASTE]: "waste",
    [CHEM_O2]: "o2",
    [CHEM_AA]: "aminoAcid",
    [CHEM_FA]: "fattyAcid",
    [CHEM_GLU]: "glucose",
    [CHEM_BIOPOLYMER]: "biopolymer",
    [CHEM_MEMBRANE]: "membrane",
    [CHEM_MARKER0]: "marker0",
  };
  for (const spec of VENT_EMISSIONS) {
    idx.push(MOLECULE_INDEX[CHEM_TO_KEY[spec.chemId]]);
  }
  return idx;
})();

const VENT_TOTAL_WEIGHT = VENT_EMISSIONS.reduce((a, s) => a + s.weight, 0);

// Compute the next dormant span (sim seconds) given the world's
// dayPeriod. Pulls one rng draw for jitter, so determinism is preserved.
export function nextVentDormantSpan(dayPeriod: number, rng: () => number): number {
  const base = VENT_CYCLE_DAYS * dayPeriod;
  const jitter = (rng() - 0.5) * 2 * VENT_CYCLE_JITTER_FRAC * base;
  return base + jitter;
}

// Construct an initial vent state. First eruption is scheduled one
// dormant span out so a fresh world is quiet at t=0.
export function makeVentState(x: number, y: number, dayPeriod: number, rng: () => number): VentState {
  return {
    x, y,
    nextEruptionAt: nextVentDormantSpan(dayPeriod, rng),
    eruptionEndsAt: 0,
    intensity: 0,
    active: false,
    emitClock: 0,
  };
}

// Pick an emission spec by weight. Walks the cumulative table; with
// only ~10 entries the linear scan is fast and keeps the ledger
// alignment trivial.
function pickEmission(rng: () => number): { spec: EmitSpec; molIdx: number } {
  const roll = rng() * VENT_TOTAL_WEIGHT;
  let acc = 0;
  for (let i = 0; i < VENT_EMISSIONS.length; i++) {
    acc += VENT_EMISSIONS[i].weight;
    if (roll < acc) return { spec: VENT_EMISSIONS[i], molIdx: VENT_EMIT_MOL_IDX[i] };
  }
  const i = VENT_EMISSIONS.length - 1;
  return { spec: VENT_EMISSIONS[i], molIdx: VENT_EMIT_MOL_IDX[i] };
}

// Tick the vent: advance the phase machine, set intensity, emit
// particles if appropriate. spawn() is the engine's pushParticle
// adapter (kept abstract so this module doesn't import sim.ts).
export function stepVent(
  world: World,
  dt: number,
  rng: () => number,
  spawn: (
    x: number, y: number, z: number,
    vx: number, vy: number, vz: number,
    r: number, chemId: number, density: number,
    molecules: Molecules,
  ) => void,
): void {
  const v = world.vent;
  if (!v) return;
  const ledger = world.ventEmitted;
  const t = world.t;

  // After the seed-ramp window closes the world is meant to be a
  // closed substrate; the vent's role was to seed abiotic chemistry +
  // localized heat during the initial fill. Permanently dormant from
  // here on: any in-progress eruption ends, no new ones scheduled.
  if (world.initialSeedDone) {
    if (v.active) {
      v.active = false;
      v.intensity = 0;
    }
    return;
  }

  // Phase transitions.
  if (!v.active && t >= v.nextEruptionAt) {
    v.active = true;
    v.eruptionEndsAt = t + VENT_WARMUP_SEC + VENT_MAIN_SEC + VENT_COOLDOWN_SEC;
    v.emitClock = 0;
  }
  if (v.active && t >= v.eruptionEndsAt) {
    v.active = false;
    v.intensity = 0;
    v.nextEruptionAt = t + nextVentDormantSpan(world.dayPeriod, rng);
    return;
  }
  if (!v.active) return;

  // Intensity envelope across the active window.
  const into = t - (v.eruptionEndsAt - (VENT_WARMUP_SEC + VENT_MAIN_SEC + VENT_COOLDOWN_SEC));
  if (into < VENT_WARMUP_SEC) {
    v.intensity = into / VENT_WARMUP_SEC;
  } else if (into < VENT_WARMUP_SEC + VENT_MAIN_SEC) {
    v.intensity = 1;
  } else {
    const c = into - VENT_WARMUP_SEC - VENT_MAIN_SEC;
    v.intensity = Math.max(0, 1 - c / VENT_COOLDOWN_SEC);
  }

  // Emit one batch per VENT_EMIT_PERIOD while active. intensity^2 keeps
  // the ramp visually punchy -- the main phase pours, warmup/cooldown
  // dribble.
  v.emitClock += dt;
  while (v.emitClock >= VENT_EMIT_PERIOD) {
    v.emitClock -= VENT_EMIT_PERIOD;
    const ramp = v.intensity * v.intensity;
    const n = Math.max(0, Math.floor(VENT_BATCH_AT_PEAK * ramp + rng()));
    for (let i = 0; i < n; i++) {
      const { spec, molIdx } = pickEmission(rng);
      const r = 1 + rng() * 1.2;
      // Spawn slightly above the vent mouth so particles enter the
      // water column rather than overlapping the rock. lateralSpread
      // gives the plume a visible width.
      const sx = v.x + (rng() - 0.5) * VENT_LATERAL_SPREAD;
      const sy = v.y - 8 - rng() * 6;
      const sz = world.depth * 0.5 + (rng() - 0.5) * VENT_DEPTH_SPREAD;
      const vy = -VENT_EXIT_SPEED * (0.6 + 0.4 * rng());
      const vx = (rng() - 0.5) * 30;
      const vz = (rng() - 0.5) * 18;
      // Tag the particle with its molecular content so dissolution +
      // ingestion paths credit the right pool. unitMass is per-particle
      // so the ledger bookkeeping is just += unitMass.
      const mols = emptyMolecules();
      const key = MOLECULE_IDS[molIdx];
      mols[key] = spec.unitMass;
      spawn(sx, sy, sz, vx, vy, vz, r, spec.chemId, spec.density, mols);
      if (ledger) ledger[molIdx] += spec.unitMass;
    }
  }
}

// Spatial temperature contribution: a soft Gaussian bubble around the
// vent that decays with distance. Reads cleanly when added to the
// baseline temperatureAt in sim.ts. Returns 0 outside ~3*radius so
// most lookups exit fast.
export function ventHeatAt(world: World, x: number, y: number): number {
  const v = world.vent;
  if (!v || !v.active) return 0;
  const dx = x - v.x;
  const dy = y - v.y;
  const d2 = dx * dx + dy * dy;
  const r2 = VENT_TEMP_RADIUS * VENT_TEMP_RADIUS;
  if (d2 > 9 * r2) return 0;
  // Gaussian: exp(-d^2 / r^2). Peak at vent, falloff to ~0.05 at 3r.
  const g = Math.exp(-d2 / r2);
  return VENT_TEMP_PEAK * v.intensity * g;
}

File: server/crates/evosim-engine/src/vent.rs
//! Hydrothermal vent. Substrate port of `src/sim/vent.ts`.
//!
//! A point feature embedded in the seafloor rock that erupts on a slow
//! schedule, emitting hot particles upward (high vy) and raising local
//! water temperature while active. The vent is a TRUE SOURCE in the
//! mass ledger -- emitted mass is recorded via the particle store, not
//! drawn from any pool, so it doesn't conserve world mass in the
//! `mass::report` sense. (TS tracks a `world.ventEmitted` ledger and
//! subtracts it from totals; we keep the cycle / emission but skip the
//! ledger until a generic-molecules pool lands.)
//!
//! Schedule (sim seconds):
//!   - Dormant for ~2 in-game days ± jitter between eruptions.
//!   - Warmup (intensity 0 -> 1) over `WARMUP_SEC`.
//!   - Main phase (intensity = 1) for `MAIN_SEC`.
//!   - Cooldown (intensity 1 -> 0) over `COOLDOWN_SEC`.
//!
//! Total active window is ~13 sim seconds.
//!
//! Materials: weighted draws from `EMISSIONS` (heavy minerals / CO2 /
//! waste dominant; abiogenic precursors AA / FA / GLU rare; MARKER0
//! rarest). One chem per particle.
//!
//! What's NOT here (kept honest):
//!   - global-saturation gate (TS blocks emissions when any vent chem
//!     hits >= 10% of total dissolved capacity) -- we don't yet have a
//!     per-chem capacity accumulator wired in
//!   - generic-molecules content tag on particles (`Molecules` map)
//!     -- our particle SoA only carries a single chem_id
//!   - vent fuel seep (a continuous bounded reduced-fuel cocktail
//!     standing pool around the vent mouth) -- belongs in chemolith
//!     and depends on the generic chems table

use crate::chem_ids::{
    CHEM_AA, CHEM_BIOPOLYMER, CHEM_CO2, CHEM_FA, CHEM_GLU, CHEM_MARKER0, CHEM_MEMBRANE, CHEM_MIN,
    CHEM_O2, CHEM_WASTE,
};
use crate::particles::{ParticleInit, ParticleStore};
use crate::rng::Mulberry32;
use serde::{Deserialize, Serialize};

/// Schedule constants. Two game days at `day_period_s=600` = 1200
/// sim seconds between eruptions; small symmetric jitter so the
/// rhythm is recognizable but not perfectly clockwork.
pub const VENT_CYCLE_DAYS: f64 = 2.0;
const VENT_CYCLE_JITTER_FRAC: f64 = 0.15;
const VENT_WARMUP_SEC: f64 = 2.0;
const VENT_MAIN_SEC: f64 = 8.0;
const VENT_COOLDOWN_SEC: f64 = 3.0;

/// Active-window total. Pre-computed so the phase machine stays
/// simple: every transition reads this once.
const VENT_ACTIVE_WINDOW: f64 = VENT_WARMUP_SEC + VENT_MAIN_SEC + VENT_COOLDOWN_SEC;

/// Emission cadence while active.
const VENT_EMIT_PERIOD: f64 = 0.10;
/// Expected particles per emission batch at peak intensity. Scales
/// by intensity^2 so warmup/cooldown shed dramatically fewer than the
/// main phase; the rare-puff design (BATCH_AT_PEAK << 1) keeps the
/// vent from dominating long-run mass.
const VENT_BATCH_AT_PEAK: f32 = 0.0004;

/// Upward emission velocity. Particles get a strong vy < 0 (UP in
/// y-down coords) + lateral / depth spread; gravity + buoyancy take
/// over once they clear the rock mouth.
const VENT_EXIT_SPEED: f32 = 90.0;
const VENT_LATERAL_SPREAD: f32 = 35.0;

/// Spatial temperature contribution: a Gaussian bubble of `VENT_TEMP_PEAK`
/// degrees inside `VENT_TEMP_RADIUS`. `vent_heat_at` reads these.
pub const VENT_TEMP_PEAK: f32 = 40.0;
pub const VENT_TEMP_RADIUS: f32 = 90.0;

/// Eruption material spec.
#[derive(Debug, Clone, Copy)]
struct EmitSpec {
    chem_id: usize,
    weight: f32,
    /// Density for the emitted particle. Vent precipitates are dense
    /// (sulfides, metal salts -> heavy minerals); organics ride on
    /// buoyancy from the upward velocity rather than density alone.
    density: f32,
}

const EMISSIONS: [EmitSpec; 10] = [
    // Common vent chemistry.
    EmitSpec { chem_id: CHEM_MIN, weight: 7.0, density: 1.8 },
    EmitSpec { chem_id: CHEM_CO2, weight: 5.0, density: 0.6 },
    EmitSpec { chem_id: CHEM_WASTE, weight: 3.0, density: 1.1 },
    // Slightly less common: dissolved O2 cycled through hot rock.
    EmitSpec { chem_id: CHEM_O2, weight: 1.5, density: 0.4 },
    // Rare: abiogenic precursors -- vents are a leading origin-of-life
    // candidate. Low weights so these read as a real find in the wild.
    EmitSpec { chem_id: CHEM_AA, weight: 1.0, density: 1.0 },
    EmitSpec { chem_id: CHEM_FA, weight: 1.0, density: 0.9 },
    EmitSpec { chem_id: CHEM_GLU, weight: 0.7, density: 1.0 },
    EmitSpec { chem_id: CHEM_BIOPOLYMER, weight: 0.6, density: 1.0 },
    EmitSpec { chem_id: CHEM_MEMBRANE, weight: 0.4, density: 0.95 },
    // Very rare: marker chemistry -- gives evolution a sniffable
    // tracer that only originates at the vent.
    EmitSpec { chem_id: CHEM_MARKER0, weight: 0.2, density: 1.0 },
];

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VentState {
    pub x: f32,
    pub y: f32,
    pub next_eruption_at: f64,
    pub eruption_ends_at: f64,
    pub intensity: f32,
    pub active: bool,
    pub emit_clock: f64,
}

impl VentState {
    /// Construct an initial vent state. First eruption is scheduled one
    /// dormant span out so a fresh world is quiet at t=0.
    pub fn new(x: f32, y: f32, day_period_s: f64, rng: &mut Mulberry32) -> Self {
        Self {
            x,
            y,
            next_eruption_at: next_dormant_span(day_period_s, rng),
            eruption_ends_at: 0.0,
            intensity: 0.0,
            active: false,
            emit_clock: 0.0,
        }
    }
}

/// Next dormant span (sim seconds) given the world's `day_period_s`.
/// One RNG draw for jitter; determinism preserved.
pub fn next_dormant_span(day_period_s: f64, rng: &mut Mulberry32) -> f64 {
    let base = VENT_CYCLE_DAYS * day_period_s;
    let jitter = (rng.next_f64() - 0.5) * 2.0 * VENT_CYCLE_JITTER_FRAC * base;
    base + jitter
}

/// Pick an emission spec by weight. Linear scan over a 10-entry table.
fn pick_emission(rng: &mut Mulberry32) -> EmitSpec {
    let total: f32 = EMISSIONS.iter().map(|s| s.weight).sum();
    let roll = rng.next_f64() as f32 * total;
    let mut acc = 0.0_f32;
    for s in &EMISSIONS {
        acc += s.weight;
        if roll < acc {
            return *s;
        }
    }
    *EMISSIONS.last().expect("EMISSIONS is non-empty")
}

/// Tick the vent: advance the phase machine, set intensity, emit
/// particles into `particles`. `t` is the world's sim seconds.
pub fn step_vent(
    vent: &mut VentState,
    particles: &mut ParticleStore,
    t: f64,
    dt: f64,
    day_period_s: f64,
    rng: &mut Mulberry32,
) {
    // Phase transitions.
    if !vent.active && t >= vent.next_eruption_at {
        vent.active = true;
        vent.eruption_ends_at = t + VENT_ACTIVE_WINDOW;
        vent.emit_clock = 0.0;
    }
    if vent.active && t >= vent.eruption_ends_at {
        vent.active = false;
        vent.intensity = 0.0;
        vent.next_eruption_at = t + next_dormant_span(day_period_s, rng);
        return;
    }
    if !vent.active {
        return;
    }
    // Intensity envelope across the active window.
    let into = (t - (vent.eruption_ends_at - VENT_ACTIVE_WINDOW)) as f32;
    let warmup = VENT_WARMUP_SEC as f32;
    let main = VENT_MAIN_SEC as f32;
    let cooldown = VENT_COOLDOWN_SEC as f32;
    vent.intensity = if into < warmup {
        into / warmup
    } else if into < warmup + main {
        1.0
    } else {
        (1.0 - (into - warmup - main) / cooldown).max(0.0)
    };
    // Emit batches.
    vent.emit_clock += dt;
    while vent.emit_clock >= VENT_EMIT_PERIOD {
        vent.emit_clock -= VENT_EMIT_PERIOD;
        let ramp = vent.intensity * vent.intensity;
        let expected = VENT_BATCH_AT_PEAK * ramp + rng.next_f64() as f32;
        let n = expected.floor().max(0.0) as usize;
        for _ in 0..n {
            let spec = pick_emission(rng);
            let r = 1.0 + rng.next_f64() as f32 * 1.2;
            let sx = vent.x + (rng.next_f64() as f32 - 0.5) * VENT_LATERAL_SPREAD;
            // Spawn slightly above the vent mouth so particles enter
            // the water column rather than overlapping rock.
            let sy = vent.y - 8.0 - rng.next_f64() as f32 * 6.0;
            let vy = -VENT_EXIT_SPEED * (0.6 + 0.4 * rng.next_f64() as f32);
            let vx = (rng.next_f64() as f32 - 0.5) * 30.0;
            particles.push(ParticleInit {
                x: sx,
                y: sy,
                vx,
                vy,
                r,
                chem_id: spec.chem_id as u8,
                density: spec.density,
                ..ParticleInit::default()
            });
        }
    }
}

/// Local temperature contribution: Gaussian bubble around the vent
/// that decays with distance. Returns 0 outside ~3 radii so most
/// lookups exit fast.
pub fn vent_heat_at(vent: &VentState, x: f32, y: f32) -> f32 {
    if !vent.active {
        return 0.0;
    }
    let dx = x - vent.x;
    let dy = y - vent.y;
    let d2 = dx * dx + dy * dy;
    let r2 = VENT_TEMP_RADIUS * VENT_TEMP_RADIUS;
    if d2 > 9.0 * r2 {
        return 0.0;
    }
    VENT_TEMP_PEAK * vent.intensity * (-d2 / r2).exp()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dormant_at_t_zero() {
        let mut rng = Mulberry32::new(1);
        let v = VentState::new(100.0, 200.0, 600.0, &mut rng);
        assert!(!v.active);
        // First eruption scheduled at t > 0, roughly one dormant span out.
        assert!(v.next_eruption_at > 0.0);
        assert!(v.next_eruption_at < VENT_CYCLE_DAYS * 600.0 * 2.0);
    }

    #[test]
    fn vent_activates_at_scheduled_time() {
        let mut rng = Mulberry32::new(1);
        let mut v = VentState::new(100.0, 200.0, 600.0, &mut rng);
        let mut ps = ParticleStore::new();
        // Step forward to just past the scheduled eruption.
        let dt = 0.1_f64;
        let mut t = 0.0_f64;
        while t < v.next_eruption_at + 0.5 {
            step_vent(&mut v, &mut ps, t, dt, 600.0, &mut rng);
            t += dt;
        }
        assert!(v.active);
    }

    #[test]
    fn vent_intensity_ramps() {
        let mut rng = Mulberry32::new(1);
        let mut v = VentState::new(100.0, 200.0, 600.0, &mut rng);
        let mut ps = ParticleStore::new();
        // Force the vent active by setting next_eruption_at to 0.
        v.next_eruption_at = 0.0;
        let dt = 0.1;
        // Step in for one warmup second: intensity should be partial.
        step_vent(&mut v, &mut ps, 0.0, dt, 600.0, &mut rng);
        step_vent(&mut v, &mut ps, 1.0, dt, 600.0, &mut rng);
        assert!(v.active);
        assert!(v.intensity > 0.0 && v.intensity <= 1.0);
        // Mid-main: intensity should be 1.
        step_vent(
            &mut v,
            &mut ps,
            VENT_WARMUP_SEC + VENT_MAIN_SEC * 0.5,
            dt,
            600.0,
            &mut rng,
        );
        assert_eq!(v.intensity, 1.0);
    }

    #[test]
    fn vent_heat_decays_with_distance() {
        let mut rng = Mulberry32::new(1);
        let mut v = VentState::new(100.0, 200.0, 600.0, &mut rng);
        v.active = true;
        v.intensity = 1.0;
        let at_vent = vent_heat_at(&v, 100.0, 200.0);
        let nearby = vent_heat_at(&v, 130.0, 200.0);
        let far = vent_heat_at(&v, 500.0, 200.0);
        assert!(at_vent > nearby);
        assert!(nearby > far);
        assert_eq!(far, 0.0, "far outside radius -> 0");
    }

    #[test]
    fn inactive_vent_emits_no_heat() {
        let mut rng = Mulberry32::new(1);
        let v = VentState::new(100.0, 200.0, 600.0, &mut rng);
        assert_eq!(vent_heat_at(&v, 100.0, 200.0), 0.0);
    }
}
